### PR TITLE
feat(resolve): add alias/fallback support

### PR DESCRIPTION
**Pull request by AdamYahid**
Adds the ability to provide alternative requests to ones being resolved.

e.g.
```ts
"alias": {
  'my-lib': 'remapped-lib', // exact by default
  'my-lib/*': 'remapped-lib/*',
  path: false
}
```

### CODE REVIEW

File: packages/resolve/src/request-resolver.ts [hunk 1]
```diff
@@ -117,6 +123,35 @@ export function createRequestResolver(options: IRequestResolverOptions): Request
     }
   }
 
+  /**
+   *
+   * @param request - the original request
+   * The function generates all the possible aliased requests, and falls back to the original request if all else failed
+   */
+  function* aliasRequestPaths(request: string) {
```
Comment (AdamYahid) [inline packages/resolve/src/request-resolver.ts:131]: Need to clean this method up a bit and add comments

File: packages/resolve/src/request-resolver.ts [hunk 2]
```diff
@@ -117,6 +123,35 @@ export function createRequestResolver(options: IRequestResolverOptions): Request
...
+    for (const [aliasedFrom, aliasedTo] of Object.entries(aliases)) {
+      if (aliasedFrom.endsWith('$') && request === aliasedFrom.slice(0, aliasedFrom.length - 1)) {
+        if (aliasedTo === false) {
+          yield false;
+        }
+        yield aliasedTo;
+      } else if (!request.endsWith('$') && request.startsWith(aliasedFrom)) {
+        if (aliasedTo === false) {
+          yield false;
+        } else if (/\.\w+$/.test(aliasedTo)) {
```
Comment (AdamYahid) [inline packages/resolve/src/request-resolver.ts:141]: @AviVahl  what do you think about bringing in path as a dependency for resolve so I can use path.extname and path.join?

File: packages/resolve/src/types.ts [hunk 3]
```diff
@@ -29,6 +32,28 @@ export interface IRequestResolverOptions {
    * If not provided, resolver will create an internal Map (still caches).
    */
   resolvedPacakgesCache?: Map<string, IResolvedPackageJson | undefined>;
+
+  /**
+   * Resolution aliases
+   * Map either a request prefix or an exact request (by adding $ at the end of it), to another request/prefix
```
Comment (AviVahl) [inline packages/resolve/src/types.ts:38]: doc is outdated. we don't use $

File: packages/resolve/src/types.ts [hunk 4]
```diff
@@ -29,6 +32,28 @@ export interface IRequestResolverOptions {
...
+   *
+   * ex.
```
Comment (AviVahl) [inline packages/resolve/src/types.ts:40]: remove this section.

File: packages/resolve/src/types.ts [hunk 5]
```diff
@@ -34,15 +33,17 @@ export interface IRequestResolverOptions {
...
 
   /**
-   * Resolution aliases and fallback
-   * Map either a request prefix or an exact request to another request/prefix
-   *
-   * ex.
-   * { xyz: '/abc/path/to/file.js', '/xyz/*': '/abc/*' } -> import 'xyz' // output /abc/path/to/file.js, import 'xyz/file' => '/abc/file.js'
-   *
-   * Aliases take precedence over other module resolutions, fallbacks attermpt if regular resolution fails
+   * Aliases for package requests.
```
Comment (AdamYahid) [inline packages/resolve/src/types.ts:36]: @AviVahl  mapping of resolved files (ie absolute paths) does not match the doc

**Pull request merged**